### PR TITLE
 Laboratory work #3, Maxim Matveev - 21FPL1

### PR DESCRIPTION
**Pull request by yourMasya**
ЪЬБ

### CODE REVIEW

File: lab_2_keywords_cooccurrence/main.py [hunk 1]
```diff
@@ -30,13 +108,19 @@ def extract_phrases(text: str) -> Optional[Sequence[str]]:
...
-                            r"|[^\s\w]+(?=$)")  # punctuation before the end of line
-    return [clean for phrase in re.split(expression, text) if (clean := phrase.strip())]
+    pattern = re.compile(r"[^\w\s]")
+    #  (?<=^)[^\s\w]+|(?<=\s)[^\s\w]+|[^\s\w]+(?=\s)|[^\s\w]+(?=$)
+
+    #  ?= -- positive lookahead     ?<= -- positive lookbehind
+    #  ?! -- negative lookahead     ?!= -- negative lookbehind
+
+    phrases = [phrase.strip() for phrase in re.split(pattern, text) if phrase]
+    while "" in phrases:
```
Comment (artyomtugaryov) [inline lab_2_keywords_cooccurrence/main.py:121]: bad implementation. It could be faster

File: lab_2_keywords_cooccurrence/main.py [hunk 2]
```diff
@@ -48,13 +132,15 @@ def extract_candidate_keyword_phrases(phrases: Sequence[str], stop_words: Sequen
...
+    if not (my_isinstance(phrases, list) and my_isinstance(stop_words, list)):
         return None
-    candidates = []
-    for phrase in [phrase.lower().split() for phrase in phrases]:
-        splits = [-1] + [index for index, word in enumerate(phrase) if word in stop_words] + [len(phrase)]
-        candidates.extend(tuple(candidate) for start, end in pairwise(splits) if (candidate := phrase[start+1:end]))
-    return candidates
+
+    text_splited = " † ".join(phrases).lower().split()
+    stop_words_destroyed = " ".join([word if word not in stop_words else "†" for word in text_splited])
```
Comment (artyomtugaryov) [inline lab_2_keywords_cooccurrence/main.py:140]: use str.replace

File: lab_2_keywords_cooccurrence/main.py [hunk 3]
```diff
@@ -48,13 +132,15 @@ def extract_candidate_keyword_phrases(phrases: Sequence[str], stop_words: Sequen
...
+    key_phrases_empty_items = [tuple(phrase.split()) for phrase in re.split("†", stop_words_destroyed)]
+    key_phrases = list(filter(lambda x: x != tuple(), key_phrases_empty_items))
```
Comment (artyomtugaryov) [inline lab_2_keywords_cooccurrence/main.py:142]: Do not use checks like `== tuple()` or `!= list()`

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Implementation for the 2 lab. The name of the PR is about 3rd

**Review by artyomtugaryov — COMMENTED** [review]
revert the implementation of the lab2

File: lab_3_keywords_textrank/main.py [hunk 4]
```diff
@@ -660,7 +620,7 @@ def get_scores(self) -> dict[int, float]:
             dict[int, float]
                 importance scores of all tokens in the encoded text
         """
-        return self._scores
+        pass
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:623]: you have to return `self._scores`

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_keywords_cooccurrence/main.py [hunk 5]
```diff
@@ -95,11 +90,9 @@ def calculate_word_scores(word_degrees: Mapping[str, int],
                           word_frequencies: Mapping[str, int]) -> Optional[Mapping[str, float]]:
     """
     Calculate the word score based on the word degree and word frequency metrics
-
```
Comment (artyomtugaryov) [inline lab_2_keywords_cooccurrence/main.py:98]: revert

File: lab_2_keywords_cooccurrence/start.py [hunk 6]
```diff
@@ -70,4 +70,4 @@ def read_target_text(file_path: Path) -> str:
 
     RESULT = UNKNOWN_PROCESSED
 
-    assert RESULT, 'Keywords are not extracted'
+    assert RESULT, 'Keywords are not extracted'
```
Comment (artyomtugaryov) [inline lab_2_keywords_cooccurrence/start.py:73]: REVERT

File: lab_3_keywords_textrank/main.py [hunk 7]
```diff
@@ -129,8 +117,8 @@ def __init__(self) -> None:
         """
         Constructs all the necessary attributes for the text encoder object
         """
-        self._word2id = {}
-        self._id2word = {}
+        self._word2id = dict()
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:120]: hm...

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
almost didn't suspect anything

File: lab_2_keywords_cooccurrence/main.py (hunk 1 above)
Comment (yourMasya) [reply lab_2_keywords_cooccurrence/main.py]: solved

File: lab_2_keywords_cooccurrence/main.py (hunk 2 above)
Comment (yourMasya) [reply lab_2_keywords_cooccurrence/main.py]: solved

File: lab_2_keywords_cooccurrence/main.py (hunk 3 above)
Comment (yourMasya) [reply lab_2_keywords_cooccurrence/main.py]: solved

File: lab_3_keywords_textrank/main.py (hunk 4 above)
Comment (yourMasya) [reply lab_3_keywords_textrank/main.py]: solved

File: lab_2_keywords_cooccurrence/main.py (hunk 5 above)
Comment (yourMasya) [reply lab_2_keywords_cooccurrence/main.py]: solved

File: lab_2_keywords_cooccurrence/start.py (hunk 6 above)
Comment (yourMasya) [reply lab_2_keywords_cooccurrence/start.py]: solved

File: lab_3_keywords_textrank/main.py (hunk 7 above)
Comment (yourMasya) [reply lab_3_keywords_textrank/main.py]: solved)0))

**Review by artyomtugaryov — APPROVED** [review]

**Pull request closed without merging**